### PR TITLE
Better argument preparation for distributions in stats package.

### PR DESCRIPTION
**Pull request by iandall**
This code improves the preparation of arguments by `argsreduce` as discussed in [#13589]. This had significant performance impact, most benchmarks are faster, some are slower and a few are dramatically (50 times) faster. This is probably needs discussion and further work. 
#### Reference issue

Addresses 13589
#### What does this implement/fix?

This is primarily a change to argsreduce, so that scalars remain scalars instead of being expanded, flattened and masked as before. 

Most distributions handle arguments in the new form by automatic broadcasting but few had to be modified to accept arguments in the new form.

#### Additional information
All tests in the stats testsuite pass.
[benchmark-my_lazywhere.txt](https://github.com/scipy/scipy/files/6042803/benchmark-my_lazywhere.txt)

### CODE REVIEW

File: scipy/stats/_distn_infrastructure.py [hunk 1]
```diff
@@ -538,20 +540,37 @@ def argsreduce(cond, *args):
...
+
+    s = cond.shape
+    # np.extract returns flattened arrays, which are not broadcastable together unless they
+    # are either the same size or size == 1.
+    ret = []
+    for arg in newargs:
+        if np.size(arg) == 1:
+            ret.append(arg)
+        else:
+            ret.append(np.extract(cond, np.broadcast_to(arg, s)))
```
Comment (tupui) [inline scipy/stats/_distn_infrastructure.py:572]: I would keep the list comprehension (with a ternary here)
```suggestion
    ret = [(arg if np.size(arg) == 1
            else np.extract(cond, np.broadcast_to(arg, s)))
           for arg in newargs]
```

**Review by rgommers — APPROVED** [review]
This is great, thanks @iandall. `stats.test()` gets sped up by about 10%, which is a pretty amazing improvement. That confirms the benchmark results - on average this is a significant improvement. It's not strange that there are some benchmarks that are up to 15% slower now; that's an okay price to pay for others getting a 50x speedup:)

**Comment by rgommers** [comment]
I pushed some minor style cleanups, it looks good to go aside from those.

**Comment by rgommers** [comment]
Looks like there's a Homebrew issue, it is building OpenBLAS and suitesparse from source. Ignoring - in it goes.

**Comment by WarrenWeckesser** [comment]
Git bisect shows that commit 2233802 is the cause of https://github.com/scipy/scipy/issues/13791.

**Pull request merged**